### PR TITLE
Reduce the amount of raw pointer the HIR API

### PR DESCRIPTION
**Pull request by P-E-P**
This draft PR is a first attempt at reducing the amount of raw pointers in the HIR API by replacing those either with references or smart pointers.

Requires #2867 to be merged first.

This PR probably requires a lot of other changes, I was trying to get something that compiles again quickly and many things may be wrong.

@philberty Is this even something you'd like to see merged ?

### CODE REVIEW

File: gcc/rust/ast/rust-ast-collector.cc [hunk 1]
```diff
@@ -2570,8 +2570,7 @@ TokenCollector::visit (LetStmt &stmt)
 {
   push (Rust::Token::make (LET, stmt.get_locus ()));
   auto &pattern = stmt.get_pattern ();
-  if (pattern)
-    visit (pattern);
+  visit (pattern);
```
Comment (P-E-P) [inline gcc/rust/ast/rust-ast-collector.cc:2573]: Need to keep the condition here.

**Comment by philberty** [comment]
I think this is a good idea, I did alot of hacky stuff in the HIR just to get stuff going but yeah this was lazy of me. If you can i would persue finishing this PR

**Comment by P-E-P** [comment]
It's been a long time since I last updated this PR. This was compiling fine but there was a lot of errors (mainly use after free and segfaults). So even if we manage to rebase it on master it'll still be a long way until merge.

**Comment by P-E-P** [comment]
This does not look good, I should have refactored the hir files before attempting to remove the raw pointers.

**Comment by P-E-P** [comment]
Putting this as ready as it will be horrible and long to review, it still requires a few fixes and a rebase though.

File: gcc/rust/checks/errors/borrowck/rust-bir-builder-pattern.h [hunk 2]
```diff
@@ -90,7 +90,7 @@ class PatternBindingBuilder : protected AbstractBuilder,
   void visit (HIR::PathInExpression &expression) override {}
   void visit (HIR::QualifiedPathInExpression &expression) override {}
   void visit (HIR::RangePattern &pattern) override {}
-};
+}; // namespace Rust
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/borrowck/rust-bir-builder-pattern.h:93]: I think this needs to be removed? probably an automatic addition by clang-format due to mismatched brackets

File: gcc/rust/checks/errors/rust-unsafe-checker.h [hunk 3]
```diff
@@ -115,7 +115,7 @@ class UnsafeChecker : public HIRFullVisitor
   virtual void visit (MatchExpr &expr) override;
   virtual void visit (AwaitExpr &expr) override;
   virtual void visit (AsyncBlockExpr &expr) override;
-  virtual void visit (InlineAsm &expr);
+  virtual void visit (InlineAsm &expr) override;
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-unsafe-checker.h:118]: I think that's part of #3244 so this needs a rebase I think?

File: gcc/rust/hir/rust-ast-lower-expr.cc [hunk 4]
```diff
@@ -511,16 +512,17 @@ ASTLoweringExpr::visit (AST::StructExprStructFields &struct_expr)
...
   delete path;
 
-  HIR::StructBase *base = nullptr;
+  tl::optional<std::unique_ptr<HIR::StructBase>> base = tl::nullopt;
   if (struct_expr.has_struct_base ())
     {
       HIR::Expr *translated_base = ASTLoweringExpr::translate (
 	struct_expr.get_struct_base ().get_base_struct ());
-      base = new HIR::StructBase (std::unique_ptr<HIR::Expr> (translated_base));
+      base = tl::optional (Rust::make_unique<StructBase> (
```
Comment (CohenArthur) [inline gcc/rust/hir/rust-ast-lower-expr.cc:520]: do we need to wrap this in a `tl::optional (` constructor call here? I though the conversion would happen automatically

File: gcc/rust/hir/rust-ast-lower-stmt.cc [hunk 5]
```diff
@@ -68,21 +69,24 @@ ASTLoweringStmt::visit (AST::LetStmt &stmt)
...
+
+  auto type
+    = stmt.has_type () ? tl::optional<std::unique_ptr<Type>> (
+	std::unique_ptr<Type> (ASTLoweringType::translate (stmt.get_type ())))
+		       : tl::nullopt;
+  auto init_expression
+    = stmt.has_init_expr ()
+	? tl::optional<std::unique_ptr<Expr>> (std::unique_ptr<HIR::Expr> (
+	  ASTLoweringExpr::translate (stmt.get_init_expr ())))
+	: tl::nullopt;
```
Comment (CohenArthur) [inline gcc/rust/hir/rust-ast-lower-stmt.cc:81]: those calls are really rough. If it works I'd switch it to this, or at least remove the `auto` and put the `tl::optional` on the left side.
```suggestion
  tl::optional<std::unique_ptr<Type>> type
    = stmt.has_type () ? std::unique_ptr<Type> (ASTLoweringType::translate (stmt.get_type ()))
		       : tl::nullopt;
  tl::optional<std::unique_ptr<HIR::Expr>> init_expression
    = stmt.has_init_expr ()
	? std::unique_ptr<HIR::Expr> (
	  ASTLoweringExpr::translate (stmt.get_init_expr ()))
	: tl::nullopt;
```

**Review by CohenArthur — APPROVED** [review]
Thank you so much for doing all that work, amazing job

File: gcc/rust/checks/errors/borrowck/rust-bir-builder-pattern.h (hunk 2 above)
Comment (P-E-P) [reply gcc/rust/checks/errors/borrowck/rust-bir-builder-pattern.h]: Rebases & fixed.

File: gcc/rust/checks/errors/rust-unsafe-checker.h (hunk 3 above)
Comment (P-E-P) [reply gcc/rust/checks/errors/rust-unsafe-checker.h]: Deleted commit.

File: gcc/rust/hir/rust-ast-lower-expr.cc (hunk 4 above)
Comment (P-E-P) [reply gcc/rust/hir/rust-ast-lower-expr.cc:520]: The compiler complains about the type size if I remove it we should therefore leave it.

File: gcc/rust/hir/rust-ast-lower-stmt.cc (hunk 5 above)
Comment (P-E-P) [reply gcc/rust/hir/rust-ast-lower-stmt.cc]: I tried but GCC 4.8 and clang on MacOS are complaining.

**Pull request merged**